### ISSUE TITLE
GraphQL over WebSocket (graphql-ws)

### ISSUE THREAD

**Issue by enisdenjo** [hasura/graphql-engine#6264]
👋 ,

I am the author of [`graphql-ws`](https://github.com/enisdenjo/graphql-ws) and its, brand new, security first, lean, [GraphQL over WebSocket Protocol](https://github.com/enisdenjo/graphql-ws/blob/master/PROTOCOL.md).

[The Protocol is aiming to become a part of the GraphQL standard with the help of the foundation's brilliant GraphQL over HTTP Work Group.](https://github.com/graphql/graphql-over-http/pull/140)

Being more than happy to help, I was about to push a PR; however, seeing the engine written in Haskell - I stepped down as I am not familiar with it. 😬

I'd like to hear your opinion on the Protocol and if you would consider supporting it in your awesome GraphQL Engine?

P.S. if you are interested on reading a short article about the lib and its inception, you can [check it out over at The Guild's blog](https://the-guild.dev/blog/graphql-over-websockets).

**Comment by chrisvariety** [comment]
👍 would love to see this. Also note per the [subscription-transport-ws Readme](https://github.com/apollographql/subscriptions-transport-ws/blob/384f9aade704c47c0dea4d77b1dd1519b6c81ebf/README.md), the `subscriptions-transport-ws` library is not being actively maintained anymore, leaving this newer protocol as the only actively maintained option.

**Comment by waterdrop01** [comment]
Would love to see this as well :)

**Comment by maxpain** [comment]
Any updates?

**Comment by alexkharech** [comment]
+1

**Comment by Hadryan** [comment]
+1

**Comment by fhattl01** [comment]
+1

**Comment by khan297** [comment]
C

**Comment by chbert** [comment]
+1

**Comment by ntziolis** [comment]
This is duly needed as much of the tooling has moves on to the new protocol.

Our use case is that we leverage `graphql-mesh` in front of hasura
- This allows for example the use of composer composition to handle business logic that gets executed before or after executing hasura resolvers
- Global renaming of resolvers and fields based on patterns (auto camelCase with a single rename)
- Extending Types / resolver results

As of right now this works great for anything but subscriptions. due to the protocol difference subscriptions cannot be passed through mesh when hasura is the datasource.

**Comment by ntziolis** [comment]
I want to stress that the following should not reduce the need for hasura to natively support the new protocol.

Since we were blocked on this issue, hence we added support in `graphql-mesh` to support the legacy protocol for downstream graphql sources:
https://github.com/Urigo/graphql-mesh/issues/2066

This means by running a minimally configured GraphQL Mesh in front of your Hausra the following is possible:
- Clients connect with new protocol to Mesh
- Mesh uses legacy protocol to communicate with hasura
- Headers are passed through from client to Hasura

If you just wane pass through all operations through, mesh configuration can be as minimal as:
```
sources: 
  - name: countries
    handler:
      graphql: 
        endpoint: https://countries.trevorblades.com/
        useWebSocketLegacyProtocol: true
```

Here is a code sandbox where you can try this with your own Hasura (simply replace the endpoint with yours):
https://codesandbox.io/s/grpahql-mesh-graphql-legacy-ws-protocol-sample-c202d

**Comment by maxpain** [comment]
Any news?

**Comment by jgoux** [comment]
> I want to stress that the following should not reduce the need for hasura to natively support the new protocol.
> 
> Since we were blocked on this issue, we added support in `graphql-mesh` to support the legacy protocol for downstream graphql sources:
> [Urigo/graphql-mesh#2066](https://github.com/Urigo/graphql-mesh/issues/2066)
> 
> This means by running a minimally configured GraphQL Mesh in front of your Hausra the following is possible:
> 
> * Clients connect with new protocol or even SSE to Mesh
> * Mesh uses legacy protocol to communicate with hasura
> * Headers are passed through from client to Hasura
> 
> If you just wane pass through all operations through, mesh configuration can be as minimal as:
> 
> ```
> sources: 
>   - name: Hasura Sample
>     handler:
>       graphql: 
>         endpoint: https://realtime-poll.hasura.app/v1/graphql
>         useWebSocketLegacyProtocol: true
> ```
> 
> Here is a code sandbox using a public hasura instance where you can try this with your own Hasura (simply replace the endpoint with yours):
> https://codesandbox.io/s/grpahql-mesh-graphql-legacy-ws-protocol-sample-c202d?file=/.meshrc.yml

Hello @ntziolis, I'm really interested by this, thank you so much for your work on url-loader.
Would you have any example when using graphql-tools/url-loader directly without graphql-mesh?

**Comment by dsandip** [comment]
Hi everyone, this is currently WIP and is actively being reviewed.

**Comment by DominikGuzei** [comment]
@dsandip is there any place we can track the progress of this integration?

**Comment by kolharsam** [comment]
Hey Folks! This has been released with the latest release of [v2.0.8](https://github.com/hasura/graphql-engine/releases/tag/v2.0.8). It is also available on Hasura Cloud too. Please feel free to open relevant issues if you do face any problems! Thanks for your patience with us!

**Comment by bitjson** [comment]
@kolharsam thank you! 

Just wanted to add a heads up here: I think the errors returned by the current Hasura implementation are malformed, so client libraries can't show errors properly: https://github.com/hasura/graphql-engine/issues/7786
